compiler/rustc_infer/src/infer/outlives/obligations.rs

```diff
@@ -500,17 +500,20 @@ where
         substs: SubstsRef<'tcx>,
         origin: infer::SubregionOrigin<'tcx>,
         region: ty::Region<'tcx>,
+        is_opaque: bool,
     ) {
         let constraint = origin.to_constraint_category();
         for k in substs {
             match k.unpack() {
                 GenericArgKind::Lifetime(lt) => {
-                    self.delegate.push_sub_region_constraint(
-                        origin.clone(),
-                        region,
-                        lt,
-                        constraint,
-                    );
+                    if !is_opaque {
+                        self.delegate.push_sub_region_constraint(
+                            origin.clone(),
+                            region,
+                            lt,
+                            constraint,
+                        );
+                    }
```

Comment:
can you add a comment here? My first reaction here was "this isn't about implying stuff, that's about adding region obligations :thinking: this has to be unsound".

So the reason this is sound is because the opaque type cannot actually directly use regions from its substs but has to instead have them as additional bounds?

Reply:
That's indeed looks not right. I guess it would incorrectly make this pass (not tested thought):
```rust
fn opaque<'a>() -> impl Captures<'a> {}
fn assert_static<T: 'static>(_: T) {}
fn test<'a>() {
    assert_static(opaque::<'a>());
}
```
I am surprised that we don't have tests that trip for this case.